### ISSUE TITLE
Performance with large number of Collapse in v4.1

### ISSUE THREAD

**Issue by chtheis** [twbs/bootstrap#26419]
I have a page with 6 collapsibles on it, each one with more than 20 in them, so I end up with about 140 collapsibles in total. The idea is to click on one of the outside collapsibles to see an overview list of collapsible items, click on one of the inner ones to see details of that item. 

Unfortunately the performance is very bad: to open one the first time takes about 10 seconds!

I dug down to the constructor of the class Collapse:
```js
        var tabToggles = $$$1(Selector.DATA_TOGGLE);

        for (var i = 0; i < tabToggles.length; i++) {
          var elem = tabToggles[i];
          var selector = Util.getSelectorFromElement(elem);

          if (selector !== null && $$$1(selector).filter(element).length > 0) {
            this._selector = selector;

            this._triggerArray.push(elem);
          }
        }
```
Nearly all time is spent in the for loop.

`tabToggles` is an array of all 140 elements with `[data-target="collapse"]`.

First question: why do I need all of them when I click on one? 

Then for each one it will get the selector and filter the list of all matching elements filter them for the element parameter. And that is where all the time is spent, almost equally distributed between `getSelectorFromElement` and `$$$1(selector).filter(element)`

`getSelectorFromElement` will select a list of matching elements and return either the selector, if some elements where found, or null if not. And the next statement in the code will do the same, then apply the filter. 

Second question: Could we save time if we had a function like `getSelectorFromElement` which could do the filter(element), too ?
I tried to pass the element to `getSelectorFromElement` and, if present, do the filtering there. This cut the time in half. But still, very slow.

**Comment by Johann-S** [comment]
Can you provide to us a CodePen ? 

And you can try that : 
```js
      var tabToggles = $.makeArray(document.querySelectorAll(Selector.DATA_TOGGLE))
        .filter(function (elem) { return !!elem.getAttribute('data-target') })

**Comment by chtheis** [comment]
I'm trying to cut it down on JSfiddle.
So far the (almost) original data are at 
https://jsfiddle.net/chtheis/8mxntvLj/

An edited version, where I try to simulate number of collapsibles and elements is at 
https://jsfiddle.net/chtheis/zou2st3e/1/

On the original, click on a time and be patient. A table will open a table with around 20 matches, each match is a <tbody>.
Click on a match and after some match details are shown,

I have other data where the matches don't have details and there opening a time is fast(er).

I think the selectors are slow. Calling them 1 time is still OK, but calling them many times in a loop makes things bad.

**Comment by chtheis** [comment]
And for your answer: I understand. And each data-toggle could have specific data-target which, in the end, could include the same element.
If so, maybe another type of optimization would be to hint if this element is supposed to be the only one or multi-target could apply.

**Comment by chtheis** [comment]
And I tried 
var tabToggles = $.makeArray(document.querySelectorAll(Selector.DATA_TOGGLE))
but it didn't make a difference. I guess the problem is not in the one time call outside the loop but in the many calls inside.

**Comment by Johann-S** [comment]
Yep maybe we should add some options to know if a collapse is multi target or not 🤔 

BTW I'll push some modification in Vanilla JS, maybe it'll be faster

**Comment by chtheis** [comment]
Am 30.04.2018 um 16:30 schrieb Johann-S:
> BTW I'll push some modification in Vanilla JS, maybe it'll be faster

I'll try them.

But note, we are not talking about 10% faster but at least 10 times 
faster :)

**Comment by chtheis** [comment]
It is much faster :)

But the mistake was on my side. I used a target like
  `data-webgen-match=137`
which worked in your old version but the new one discovered that is an invalid syntax and instead should read
  `data-webgen-match="137"`

I updated my small [Fiddle](https://jsfiddle.net/chtheis/zou2st3e/2/)
In the beginning of createInner you see both syntax, the wrong one and the correct one.
I don't know (yet) why the wrong syntax slows down so much, are there exceptions thrown and `caught?`

**Comment by chtheis** [comment]
Update: I found out that without the quotes it is not a valid CSS selector and thus will use jQuery selector engine (which does support it) instead of the native browser API.

**Comment by herbalite** [comment]
There is room for improvement in util.getSelectorFromElement()

Look at this line
 
`const $selector = $(document).find(selector)`

Without $(document) being cached between calls of the function, in essence jQuery() is called twice. And even if $(document) is being cached, the results of find() are not inspiring. With ideas and loops from the follwing old conversation at https://stackoverflow.com/questions/1854859/jquery-performance-wise-what-is-faster-getelementbyid-or-jquery-selector

and the code provided in this thread to build the collapsibles, we can see some interesting stats. 

The loop is like this on my environment

`var j=0;
 for(i=0; i < 1000000; i++)
 {
   if ($(document).find('#outer-9').length > 0)
   {
     j++;
   }
}`
This takes 758 millseconds


Introducing caching
`var doc = $(document); 
...
doc.find('#outer-9').length > 0`

gets down to 562 milliseconds.


Can we do better? Yes. The next 2 ways are probably not feasible, as it depends on the data-target being an ID
`$('#outer-9').length > 0`

takes 282 milliseconds. Not bad, but we can do better in jQuery?


Yes. Passing jQuery an (DOM) Element is almost twice as fast as passing a string

`$(document.getElementById('#outer-9')).length > 0`

this takes 146 milliseconds. If speed is king then that code right above is a good starting point.


To stay with the ability to use other lookup methods the following seems so far the best comprise

`$(document.querySelector('#outer-9')).length > 0`

This takes 214 milliseconds. Overall the best solution IMO.  More than 3 times faster than it is right now


If things have to rely on 

`$(document.querySelectorAll('#outer-9')).length > 0`

however, that takes 1222 milliseconds

Not good. However should collapsible have multiple targets?


So template developers might want to prefer using id's and 

Bootstrap changing that line in util.getSelectorFromElement() to

`const $selector = $(document.querySelector(selector))`

might be an improvement as it is compatible down to IE8. And if there are similar uses throughout Bootstrap there might be gained even more.

**Comment by Johann-S** [comment]
Fixed by : #26422
